File: examples/utils_auth.js
// Auth utilities — JavaScript module A
const crypto = require('crypto');

function hashPassword(password, salt = null) {
    if (salt === null) {
        salt = crypto.randomBytes(32).toString('hex');
    }
    const hash = crypto.pbkdf2Sync(password, salt, 100000, 64, 'sha256');
    return { hash: hash.toString('hex'), salt };
}

function verifyPassword(password, storedHash, salt) {
    const { hash } = hashPassword(password, salt);
    return crypto.timingSafeEqual(Buffer.from(hash), Buffer.from(storedHash));
}

function generateToken(length = 32) {
    return crypto.randomBytes(length).toString('hex');
}

function validateEmail(email) {
    const pattern = /^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$/;
    return pattern.test(email);
}

function sanitizeUsername(username) {
    return username.replace(/[^a-zA-Z0-9_.-]/g, '');
}

module.exports = { hashPassword, verifyPassword, generateToken, validateEmail, sanitizeUsername };

File: examples/utils_auth.py
"""Auth utilities — module A. This represents existing code in a repo."""

import hashlib
import hmac
import os


def hash_password(password: str, salt: str | None = None) -> tuple[str, str]:
    """Hash a password with a salt using SHA-256."""
    if salt is None:
        salt = os.urandom(32).hex()
    hashed = hashlib.pbkdf2_hmac("sha256", password.encode(), salt.encode(), 100000)
    return hashed.hex(), salt


def verify_password(password: str, hashed: str, salt: str) -> bool:
    """Verify a password against a stored hash."""
    computed, _ = hash_password(password, salt)
    return hmac.compare_digest(computed, hashed)


def generate_token(length: int = 32) -> str:
    """Generate a random token for session management."""
    return os.urandom(length).hex()


def validate_email(email: str) -> bool:
    """Check if an email address is roughly valid."""
    import re
    pattern = r"^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$"
    return bool(re.match(pattern, email))


def sanitize_username(username: str) -> str:
    """Remove dangerous characters from a username."""
    import re
    return re.sub(r"[^a-zA-Z0-9_.-]", "", username)
